### PR TITLE
feat(impr): Vercel AI SDK Update

### PR DESCRIPTION
**Pull request by Bran18**
# AI SDK Migration and Update to Version 3.3

## Summary
This PR updates our AI integration to use the latest Vercel AI SDK (version 3.3), focusing on improving our handling of OpenAI, Anthropic, and Perplexity AI models. The changes primarily affect our `actions.tsx` - `ai-helpers` - `route.tsx` - `chat.tsx` files and related types.

## Reasoning
The migration to the latest AI SDK version allows us to leverage new features, improve performance, and maintain consistency with the latest best practices for AI integration. The use of provider-specific SDK modules (`@ai-sdk/openai`, `@ai-sdk/anthropic`) allows for more tailored interactions with each AI service.

## Documentation used: 
- [Migrate 3.0 to 3.1](https://sdk.vercel.ai/docs/troubleshooting/migration-guide/migration-guide-3-1)
- [Migrate 3.1 to 3.2](https://sdk.vercel.ai/docs/troubleshooting/migration-guide/migration-guide-3-2#migration-guide-31-to-32)
- [Migrate 3.2 to 3.3](https://sdk.vercel.ai/docs/troubleshooting/migration-guide/migration-guide-3-3#migration-guide-32-to-33)

## Key Changes

1. **Updated AI SDK Dependencies**:
   - Upgraded from version 2.2.25 to [3.3.0](https://vercel.com/blog/vercel-ai-sdk-3-3-3OnRtxG6a0rwvcJVu3qADv) of the Vercel AI SDK.
   - Added `@ai-sdk/openai` and `@ai-sdk/anthropic` [provider](https://sdk.vercel.ai/docs/foundations/providers-and-models) packages.

2. **Refactored `actions.tsx`**:
   - Replaced direct API calls with SDK provider instances for OpenAI and Anthropic.
   - Updated the `createResponseStream` function to use the new [streamText](https://sdk.vercel.ai/docs/ai-sdk-core/generating-text#streamtext) method from the AI SDK. 
   - Implemented a custom `convertToCoreMessages` function to ensure consistent message formatting across providers.

> Why do we need a custom `convertToCoreMessages`: Our custom implementation **allows** us to handle specific cases, such as complex content structures or additional roles, ensuring that our application logic remains intact. And the SDK's [convertToCoreMessages( ) ](https://sdk.vercel.ai/docs/reference/ai-sdk-ui/convert-to-core-messages#converttocoremessages)expects a different message format (ConvertibleMessage[]) than what we use (ChatCompletionMessageParam[] | MessageParam[]). Specifically, the content property in our messages can be more complex (e.g., an array of ChatCompletionContentPartText[]), which is not supported by the SDK's function. 

3. **Anthropic Integration**:
   - Integrated the Anthropic provider using `createAnthropic` from `@ai-sdk/anthropic`.
   - Enabled cache control for Anthropic models.

4. **OpenAI and Perplexity Updates**:
   - Refactored OpenAI integration to use the new SDK provider.
   - Updated Perplexity integration to align with the new SDK structure.

5. **Type Updates**:
   - Updated types in `types.ts` to align with the new SDK version.
   - Added `CoreMessage` type usage for improved type safety.

### Potential Impacts

- New improved capabilities response formats, structured data or streaming behavior due to SDK updates.

### Known issues: 

1. [Common Issues](https://sdk.vercel.ai/docs/troubleshooting/common-issues#common-issues)

## Future Improvements

- Explore and implement additional features provided by the new SDK version.

## Additional Notes
- This update maintains backwards compatibility with our existing message formats.
- The custom `convertToCoreMessages` function may need future updates if provider-specific message formats change.

## Summary by CodeRabbit

- **New Features**
	- Introduced a new AI model option "GPT-4o" in the chat component.
	- Added a new function for converting messages to a structured format.
  
- **Improvements**
	- Enhanced error handling and validation in the chat API, improving robustness.
	- Streamlined initialization of AI clients, simplifying setup and error management.
	- Updated type definitions to support more detailed configurations for AI interactions.

- **Dependency Updates**
	- Updated dependencies for improved functionality and compatibility with AI SDKs.

### CODE REVIEW

**Review by AndlerRL — APPROVED** [review]
LGTM! The changes are handling the cases right for each model that we have setup, so as the error control on the APIs, which would be really helpful with the issue that we face on uncontrolled errors when Ai fails, we only need to double check if we are sending this error message to the client.

I only found a additional folder that was added and a small improvement. After that, we are good to go with this PR! :+1:

File: apps/masterbots.ai/lib/ai-helpers.tsx [hunk 1]
```diff
@@ -64,10 +64,26 @@ export function setStreamerPayload(
...
     default:
-      return payload
+      return payload;
   }
 }
+
+//* Convert an array of messages into the CoreMessage format
+export function convertToCoreMessages(messages: ChatCompletionMessageParam[]): CoreMessage[] {
+  return messages.map(msg => {
+    switch (msg.role) {
```
[the remark this answers is not in the document]
Comment (Bran18) [reply apps/masterbots.ai/lib/ai-helpers.tsx]: Yeah, agree. And the other ones are just: tool and function.

**Review by AndlerRL — APPROVED** [review]
LGTM! Nice upgrade here, now without fear we can create our own Custom Provider for WordWare :grin: Merging!

@AnoukRImola -- can you collaborate on do few testing of this upgrade? To create chats and see if it is behaving as it should (creating new chats, continuing the chat).

**Comment by AndlerRL** [comment]
[Update to VercelAi SDK 3.3](https://trello.com/c/miVTrsDy/118-update-to-vercelai-sdk-33)

**Pull request merged**